Feature/bbmc

nonameletters opened the pull request.

alatarum commented on commands/host.vegman:
```diff
@@ -224,6 +224,55 @@ function cmd_virtualmedia_umount {
   usb-ctrl eject "usb.ms.cli"
 }
 
+# @restrict cmd_config tech,admin
```
don't mention tech user here, it will be added by recipe later.

nest1ing commented on commands/host.vegman:
```diff
@@ -224,6 +224,55 @@ function cmd_virtualmedia_umount {
+# @doc cmd_config_status
+# Show boot settings
+function cmd_config_status {
+    ipmitool chassis bootparam get 5
```
I guess that `expect_noarg "$@"` should be at the first line.

nest1ing commented on commands/host.vegman:
```diff
@@ -224,6 +224,55 @@ function cmd_virtualmedia_umount {
+  subcommand "$@"
+}
+
+# @sudo cmd_config_default tech,admin
```
Remove `tech` user mentioning as it's for internal using only and will be added by yocto.

nest1ing commented on commands/host.vegman:
```diff
@@ -224,6 +224,55 @@ function cmd_virtualmedia_umount {
+# @sudo cmd_config_default tech,admin
+# @restrict cmd_config_default tech,admin
+# @doc cmd_config_default
+# Restore the factory default Host config
```
If the confirmation is really required, describe it here

alatarum commented on commands/host.vegman:
```diff
@@ -224,6 +224,53 @@ function cmd_virtualmedia_umount {
+# @restrict cmd_config_nodefault tech,admin
+# @doc cmd_config_nodefault
+# Undo if the factory default Host was set
+function cmd_config_nodefault {
```
I would make this function and status as subcommands (options) of cmd_config_default, like:
> host config default --cancel
> host config default --status

nest1ing commented on commands/host.vegman:
```diff
@@ -224,6 +224,53 @@ function cmd_virtualmedia_umount {
+# @doc cmd_config_nodefault
+# Undo if the factory default Host was set
+function cmd_config_nodefault {
+  echo "Undo Host BIOS setting reset"
```
I assume the confirmation is also required here.
If not - add `expect_noarg "$@"` at the first line.

alatarum commented on commands/host.vegman:
```diff
@@ -224,6 +224,53 @@ function cmd_virtualmedia_umount {
+function cmd_config_nodefault {
+  echo "Undo Host BIOS setting reset"
+  
+  ipmitool chassis bootdev none >/dev/null
```
I'm bit confused, why you use confirmation to set flag, but don't use to reset it... But maybe it have sence

alatarum commented on commands/host.vegman:
```diff
@@ -224,6 +224,53 @@ function cmd_virtualmedia_umount {
+# @doc cmd_config_status
+# Show boot settings
+function cmd_config_status {
+    ipmitool chassis bootparam get 5
```
Shouldn't you process the answer? How it looks like?

nonameletters replied:
Boot parameter version: 1
Boot parameter 5 is valid/unlocked
Boot parameter data: 8000000000
 Boot Flags :
   - Boot Flag Valid
   - Options apply to only next boot
   - BIOS PC Compatible (legacy) boot 
   - Boot Device Selector : No override
   - BIOS verbosity : System Default
   - Console Redirection control : Console redirection occurs per BIOS configuration setting (default)
   - BIOS Mux Control Override : BIOS uses recommended setting of the mux at the end of POST

Commit:
host: Add 'config default' command

Add a command to request resetting BIOS settings to default
or to cancel the previously given request.

End-user-impact: Add next commands to cli:
                 - `host config default [-y | --yes]`
                   During next boot UEFI resets host to default state
                 - `host config default --cancel [-y | --yes]`
                   Cancel the previously given request if not yet done
                 - `host config default --status [-y | --yes]`
                   Print the UEFI settings reset request state

Signed-off-by: Vladimir Mitrofanov <v.mitrofanov@yadro.com>

nest1ing commented on commands/host.vegman:
```diff
@@ -224,6 +224,84 @@ function cmd_virtualmedia_umount {
+
+  while [[ $# -gt 0 ]]; do
+    case "${1}" in
+      --cancel) loc_config_default_cancel;;
```
`--yes` will never be passed to the function.
As a result the confirmation will be asked in any case.

nest1ing commented on commands/host.vegman:
```diff
@@ -224,6 +224,84 @@ function cmd_virtualmedia_umount {
+# @sudo cmd_config_default admin,operator,user
+# @doc cmd_config_default
+# Host set/reset UEFI boot configuration on next boot.
+# Possible flags: --cancel, --status
```
I assume the options should be explained more detailed.
And `--yes` also should present here.
This comment is used as a help message shown by `host config default --help`.

AlexanderAmelkin commented on commands/host.vegman:
```diff
@@ -224,6 +224,86 @@ function cmd_virtualmedia_umount {
 }
 
+
+function loc_config_default_set {
```
No need to call them `loc_*`. Anything that doesn't start with `cmd_` is safe. These scripts are anyway sourced into `clicmd`.

AlexanderAmelkin commented on commands/host.vegman:
```diff
@@ -224,6 +224,86 @@ function cmd_virtualmedia_umount {
+
+  echo "Cancel reset to default host settings on next boot."
+  if [[ -z "${yes}" ]]; then
+    confirm
```
I guess we also need to state here undert this if that "This operation will also clear the boot device settings, even if the request to clear settings was not actually in effect".

Or we could use `busctl` below to only reset the clear-cmos option without affecting the other boot settings.

AlexanderAmelkin commented on commands/host.vegman:
```diff
@@ -224,6 +224,86 @@ function cmd_virtualmedia_umount {
+  val=${strArr[3]}
+  len=${#val}
+
+  if [[ $((16#${val:$len-8:1})) -ge 8 ]]; then
```
Why all this magic? Why not just grep `ipmitool` output for `CMOS Clear` ? As far as I can see, that's what `ipmitool` would pring if the flag is set:

```c
if(rsp->data[3] & BF2_CMOS_CLEAR)                                 
        printf("   - CMOS Clear\n"); 
```

AlexanderAmelkin replied:
You're still doing magic. Please either explain why, or switch to checking for "CMOS Clear" in the output of `ipmitool`.

AlexanderAmelkin commented on commands/host.vegman:
```diff
@@ -224,6 +224,86 @@ function cmd_virtualmedia_umount {
+  if [[ $# -gt 0 ]]; then
+    case "${1}" in
+      --cancel)
+                  if [[ $# -ge 2 ]]; then
```
Something is wrong with indentation here. Please do it like this:
```
function...
  if...
    case...
      --cancel)
        cancel_code
        ;;
      --status)
        status_code
        ;;
      -y | --yes)
        yes_code
        ;;
      *) echo ...
...

AlexanderAmelkin commented on commands/host.vegman:
```diff
@@ -224,6 +224,86 @@ function cmd_virtualmedia_umount {
+                    loc_config_default_cancel
+                  fi;;
+      --status)   loc_config_default_status;;
+      -y | --yes) loc_config_default_set ${1};;
```
Why? Can't I use `--yes` with `--cancel` ? That's strange.
I think you need to move `--yes` processing out of the set and cancel functions.
You could determine the name of the function to call in this option processing loop, and then just call it with just a single `${yes}` argument.

AlexanderAmelkin commented on commands/host.vegman:
```diff
@@ -224,6 +224,86 @@ function cmd_virtualmedia_umount {
+# @sudo cmd_config_default admin,operator,user
+# @doc cmd_config_default
+# Host set/reset UEFI boot configuration on next boot.
+# Possible flags: --cancel, --status, --yes to approve without prompt
```
This doesn't look like options help in other commands. Take a look, for example, at `cmd_info_version` in `bmc.vegman`:

```
# @doc cmd_info_version
# Show version of BMC and BIOS/UEFI
#   -b, --bmc   Show BMC version
#   -u, --bios  Show BIOS/UEFI version
```

AlexanderAmelkin commented on commands/host.vegman:
```diff
@@ -224,6 +224,86 @@ function cmd_virtualmedia_umount {
+  subcommand "$@"
+}
+
+# @sudo cmd_config_default admin,operator,user
```
A `@restrict` clause is definitely missing. The `@sudo` just tells the script that it must creade a sudoers entry, `@restrict` actually restricts the users from seeing and executing the command.

Also, both `@restrict` and `@sudo` must contain just `admin` for this command.

alatarum commented on commands/host.vegman:
```diff
@@ -224,6 +224,96 @@ function cmd_virtualmedia_umount {
+#  --status  - Shows will be or not UEFI reset during next boot
+#  -y, --yes - does not ask for interactive confirmation
+function cmd_config_default {
+  if [[ $# -gt 0 ]]; then
```
I suggest to write something like 
```
  local yes=""
  local action="set"
  while [[ $# -gt 0 ]]; do
    case "$1" in
      --cancel) action="cancel";;
      --status) action="status";;
      -y | --yes) yes="y";;
      *) abort_badarg "$1";;
    esac
    shift
  done
```
and then call function depend on ${action} value

AlexanderAmelkin replied:
Exactly. That's what I suggested earlier.

AlexanderAmelkin replied:
Hmm... I somehow don't see my comment where I suggested exactly that.

AlexanderAmelkin replied:
Please do as Andrey suggests here and I suggested in the previous review round.

AlexanderAmelkin commented on commands/host.vegman:
```diff
@@ -224,6 +224,96 @@ function cmd_virtualmedia_umount {
+# Host set/reset UEFI boot configuration on next boot.
+#  --cancel  - Undo reset UEFI boot configuration, if it was set by mistake
+#  --status  - Shows will be or not UEFI reset during next boot
+#  -y, --yes - does not ask for interactive confirmation
```
"does not" -> "Do not"

AlexanderAmelkin commented on commands/host.vegman:
```diff
@@ -224,6 +224,90 @@ function cmd_virtualmedia_umount {
+
+function config_default_set {
+  local yes=""
+  while [[ $# -gt 0 ]]; do
+    case "$1" in
+      -y | --yes) yes="y";;
+      *) abort_badarg "$1";;
+    esac
+    shift
+  done
```
You don't need this anymore. You either pass `-y` in `$1` (that is, a non-empty `$1`), or you don't pass anything (empty `$1`). You only need this:

```
local yes="${1-}"

if [[ -z "${yes}" ]]; then
...
```

That was the whole idea behind determining the action in the main function that you don't have to repeat the code here and in cancel function.

AlexanderAmelkin commented on commands/host.vegman:
```diff
@@ -224,6 +224,90 @@ function cmd_virtualmedia_umount {
+
+function config_default_cancel {
+  local yes=""
+  while [[ $# -gt 0 ]]; do
+    case "$1" in
+      -y | --yes) yes="y";;
+      *) abort_badarg "$1";;
+    esac
+    shift
+  done
```
same here

AlexanderAmelkin commented on commands/host.vegman:
```diff
@@ -224,6 +224,90 @@ function cmd_virtualmedia_umount {
+# @sudo cmd_config_default admin
+# @restrict cmd_config_default admin
+# @doc cmd_config_default
+# Host set/reset UEFI boot configuration on next boot.
```
Change to "Reset UEFI BIOS configuration to defaults on the next boot or cancel that request"

AlexanderAmelkin commented on commands/host.vegman:
```diff
@@ -205,7 +205,7 @@ function cmd_virtualmedia_mount {
 # @doc cmd_virtualmedia_umount
 # Unmount the virtual media
-#  -y, --yes  - does not ask for interactive confirmation
+#  -y, --yes  - do not ask for interactive confirmation
```
While you're at it, please change it to capitalized 'Do'. Also, please make this refactoring a separate commit (and please check in other files where `--yes` is specified. Let's make the messge identical everywhere).

AlexanderAmelkin commented on commands/host.vegman:
```diff
@@ -224,6 +224,90 @@ function cmd_virtualmedia_umount {
+# @doc cmd_config_default
+# Host set/reset UEFI boot configuration on next boot.
+#  --cancel  - Undo reset UEFI boot configuration, if it was set by mistake
+#  --status  - Shows will be or not UEFI reset during next boot
```
Change to "Show the status of the settings reset request"

AlexanderAmelkin commented on commands/host.vegman:
```diff
@@ -224,6 +224,90 @@ function cmd_virtualmedia_umount {
+# @restrict cmd_config_default admin
+# @doc cmd_config_default
+# Host set/reset UEFI boot configuration on next boot.
+#  --cancel  - Undo reset UEFI boot configuration, if it was set by mistake
```
Change to "Cancel the settings reset request if it hasn't been processed yet"

AlexanderAmelkin commented on commands/host.vegman:
```diff
@@ -224,6 +224,90 @@ function cmd_virtualmedia_umount {
+    cancel) config_default_cancel $yes;;
+    status) config_default_status;;
+    set)    config_default_set $yes;;
+    *) echo "Unsupported parameter. Use --status, --cancel or --yes to
+approve";;
```
Please don't indent like this. Close the quote and use line continuation `\` to do proper indentation:

```
echo "Long string starts here" \
     "and continues on the next line" \
     "with proper indentation."
```

nest1ing commented on commands/host.vegman:
```diff
@@ -224,6 +224,90 @@ function cmd_virtualmedia_umount {
+
+function config_default_set {
+  local yes=""
+  while [[ $# -gt 0 ]]; do
```
I guess that this code is excess now as the `--yes` is already checked by the caller and passed to here only when it really specified by the user.
So, it seems to me, that `[[ -z "${1:-}" ]]` is enough to determine the confirmation is needed.

nest1ing commented on commands/host.vegman:
```diff
@@ -224,6 +224,90 @@ function cmd_virtualmedia_umount {
+    confirm
+  fi
+
+  ipmitool chassis bootdev none clear-cmos=yes >/dev/null
```
I suggest to use `busctl` here too, as it would be a shorter way.

AlexanderAmelkin replied:
`ipmitool` here also clears the other boot device settings, and I can't say that's bad for the 'reset to defaults' case.

nest1ing replied:
Ok, let it be as is.

nest1ing commented on commands/host.vegman:
```diff
@@ -224,6 +224,90 @@ function cmd_virtualmedia_umount {
+
+function config_default_status {
+  expect_noarg "$@"
+  if (ipmitool chassis bootparam get 5 | grep -i 'CMOS Clear'>/dev/null); then
```
Probably `busctl` would also be better here.

AlexanderAmelkin replied:
I'm fine with `ipmitool` here. It gives a more precise status because that's essentially what the BIOS will do.

nest1ing replied:
Ok, as you believe that it's reasonable as let it be as is.

nest1ing commented on commands/host.vegman:
```diff
@@ -224,6 +224,90 @@ function cmd_virtualmedia_umount {
+
+function config_default_cancel {
+  local yes=""
+  while [[ $# -gt 0 ]]; do
```
This code is excess too, by the same reason.

Commit:
Refactor help formatting

- Capitialize the beginning of help messages
- Make formatting uniform (indents, separators, etc.)
- Fix some language

Signed-off-by: Alexander Amelkin <a.amelkin@yadro.com>
Signed-off-by: Vladimir Mitrofanov <v.mitrofanov@yadro.com>

alatarum commented:
> maybe use `abort_badarg "$action"` here

It's not "bad arg", you normally can't get here because of value for action assigned in "case" above, arguments are not passed to action directly.
But we can use this helper just for simplicity.

AlexanderAmelkin commented on commands/host.vegman:
```diff
@@ -303,8 +289,8 @@ function cmd_config_default {
-    *) echo "Unsupported parameter. Use --status, --cancel or --yes to
-approve";;
+    *) echo "Unsupported parameter. Use --status, \
+             --cancel or --yes to approve";;
```
Wrong. This will print a long string of spaces between "--status," and "--cancel".
You need to close the quote before `\` and then reopen it after indentaion. See my previous comment.

Also, why are there 3 commits now? I expected 2: one for the config_default, and one for help refactoring.

AlexanderAmelkin replied:
Oh, and I totally missed that! Andrey was correct. You can never get to this line at all!
You've already directly set `action` in the above loop to one of the three values.
I'd remove this `case` completely and just call `config_default_${action} "${yes}"` here.
